ray_on_golem/server/cluster/nodes.py

```diff
@@ -435,13 +439,47 @@ async def _on_monitor_check_failed(
 class HeadClusterNode(WorkerClusterNode):
     """Self-contained element that represents explicitly a Ray head node."""
 
-    webserver_port: int
+    _webserver_port: int
+    _ray_gcs_expose_port: Optional[int]
+    _ray_dashboard_expose_port: Optional[int]
+
+    def __init__(
+        self,
+        webserver_port: int,
+        ray_gcs_expose_port: Optional[int],
+        ray_dashboard_expose_port: Optional[int],
+        **kwargs,
+    ) -> None:
+        self._webserver_port = webserver_port
+        self._ray_gcs_expose_port = ray_gcs_expose_port
+        self._ray_dashboard_expose_port = ray_dashboard_expose_port
+
+        # Late init as it will trigger `_prepare_sidecars` that requires fields from above
+        super().__init__(**kwargs)
 
     def _prepare_sidecars(self) -> Collection[ClusterNodeSidecar]:
-        return (
+        sidecars = [
             *super()._prepare_sidecars(),
-            HeadNodeToWebserverTunnelClusterNodeSidecar(self),
-        )
+            PortTunnelClusterNodeSidecar(node=self, local_port=self._webserver_port, reverse=True),
+        ]
+
+        if self._ray_gcs_expose_port is not None:
```

Comment:
why `is not None`? you can't open port 0 anyway ... 

```suggestion
        if self._ray_gcs_expose_port:
```